tests/src/test/java/alluxio/client/rest/S3ClientRestApiTest.java

```diff
@@ -744,6 +748,35 @@ private void putObjectTest(String bucket, String objectKey, byte[] object, Strin
             S3Constants.XATTR_STR_CHARSET));
   }
 
+  @Test
+  public void testGetDeletedObject() throws Exception {
+    String bucket = "bucket";
+    String objectKey = "object";
+    String object = CommonUtils.randomAlphaNumString(DATA_SIZE);
+    final String fullObjectKey = bucket + AlluxioURI.SEPARATOR + objectKey;
+    AlluxioURI bucketURI = new AlluxioURI(AlluxioURI.SEPARATOR + bucket);
+    AlluxioURI objectURI = new AlluxioURI(AlluxioURI.SEPARATOR + fullObjectKey);
+
+    createBucketRestCall(bucket);
+    createObject(fullObjectKey, object.getBytes(), null, null);
+
+    // free the object in alluxio and delete it in UFS.
+    mFileSystemMaster.free(objectURI,
+        FreeContext.mergeFrom(FreePOptions.newBuilder().setForced(true)));
+    FileUtils.deleteQuietly(
+        new File(sResource.get().getAlluxioHome() + "/underFSStorage/" + fullObjectKey));
+
+    // Verify the object is exist in the alluxio.
+    List<FileInfo> fileInfos =
+        mFileSystemMaster.listStatus(bucketURI, ListStatusContext.defaults());
+    Assert.assertEquals(1, fileInfos.size());
+    Assert.assertEquals(objectURI.getPath(), fileInfos.get(0).getPath());
+
+    // Verify 404 status will be returned by Getting Object
+    HttpURLConnection connection = getObjectRestCallWithError(fullObjectKey);
+    Assert.assertEquals(404, connection.getResponseCode());
```

Comment:
I'd like to verify the response body XML as well; making sure that the corresponding error code and other fields are there.

Reply:
Actualy the code is like this in `HttpURLConnetion`, so if it's a 404 error, I try to read the stream and I can only get the FileNotFoundException
`               if (respCode >= 400) {
                    if (respCode == 404 || respCode == 410) {
                        throw new FileNotFoundException(url.toString());
                    } else {
                        throw new java.io.IOException("Server returned HTTP" +
                              " response code: " + respCode + " for URL: " +
                              url.toString());
                    }
                }`

Reply:
Sorry, I'm wrong. I need to use getErrorStream to read the error response and I have verified the response body in the latest commit.

Reply:
Ah sorry, this is a little hard to explain without examples so I'll provide that here:

We create an `S3Error` object and serialize it as a string in the HTTP `Response` body (eg: [here](https://github.com/Alluxio/alluxio/pull/15927/files#r928517355)). Notice that `S3Error errorResponse = new S3Error(e.getMessage(), s3ErrorCode);`
```
<Error><RequestId></RequestId><Code>NoSuchKey</Code><Resource>Failed to open key: autobots-test/tmp/testdata.txt bucket: bucket-yuyang-test1 attempts: 2 error: The specified key does not exist. (...)</Resource></Error>
```
- The `<Resource>` tag is `e.getMessage()`. What we want ideally is for that tag to have the Object key (`autobots-test/tmp/testdata.txt`)
- It would look like this:
```
s3ErrorCode.setMessage(e.getMessage());
S3Error errorResponse = new S3Error(<S3 Object/Bucket name>, s3ErrorCode);
```


The code to check for the intended output format would be like this:
```
HttpURLConnection connection = getObjectRestCallWithError(fullObjectKey);
Assert.assertEquals(404, connection.getResponseCode());
// Parse S3Error XML from response body
S3Error response = new XmlMapper().readerFor(S3Error.class).readValue(connection.getErrorStream());
Assert.assertEquals(response.getResource(), objectKey);
Assert.assertEquals(response.getCode(), S3ErrorCode.Name.NO_SUCH_KEY);
```

I believe this was already a problem before when it came to handling exceptions, so it's not new to your PR. If you could refactor `S3RestUtils.createErrorResponse()` to require a `String resource` parameter and use that instead of `e.getMessage()` for the `S3Error` then we should do that.